File: sinchai/mcp_server.py
import json
import os

from mcp.server.mcpserver import MCPServer

from sinchai import clock, config, db, engine, ledger, reports

mcp = MCPServer("sinchai")
_cfg = None
_db_path = None


def _con():
    return db.open_db(_db_path, query_only=True)


def _ctrl(con):
    row = db.get_controller_row(con)
    if row is None:
        return False, clock.to_iso(clock.now()), True
    age = (clock.now() - clock.from_iso(row["heartbeat_at"])).total_seconds()
    return age < 15, row["clock_now"], row["source"] == "sim" or row["demo"] == 1


def _base(con):
    running, clock_now, simulated = _ctrl(con)
    age_min = None
    if clock_now:
        age_min = round((clock.from_iso(clock.to_iso(clock.now())) - clock.from_iso(clock_now)).total_seconds() / 60.0, 1)
    return {"as_of": clock_now, "data_age_minutes": age_min, "controller_running": running, "simulated": simulated}


@mcp.tool(description="List all zones with crop, moisture, valve state and data age.")
def list_zones():
    con = _con()
    result = _base(con)
    result["zones"] = []
    for zone in db.get_zones(con):
        r = db.get_recent_readings(con, zone["id"], 1)
        m = r[0]["moisture_pct"] if r else None
        age = round((clock.from_iso(clock.to_iso(clock.now())) - clock.from_iso(r[0]["ts"])).total_seconds() / 60.0, 1) if r else None
        result["zones"].append({"id": zone["id"], "name": zone["name"], "crop": zone["crop"],
                                 "moisture_pct": m, "reading_age_minutes": age, "valve_open": bool(zone["valve_open"])})
    con.close()
    return result


@mcp.tool(description="Get status and recommendation for one zone. zone_id must match a real zone.")
def get_zone_status(zone_id: int):
    con = _con()
    zone = next((z for z in db.get_zones(con) if z["id"] == zone_id), None)
    if zone is None:
        con.close()
        return {"error": f"zone {zone_id} not found"}
    crop = db.get_crop(con, zone["crop"])
    readings = db.get_recent_readings(con, zone_id, 24)
    now_iso = clock.to_iso(clock.now())
    rec = engine.decide(zone, crop, readings, None, now_iso,
                        clock.local_minutes(clock.now(), _cfg["location"]["utc_offset_minutes"]), _cfg)
    open_alerts = [{"kind": a["kind"], "message": a["message"]}
                   for a in db.get_open_alerts(con) if a["zone_id"] == zone_id]
    result = _base(con)
    result.update({"zone": zone["name"], "crop": zone["crop"],
                   "moisture_pct": readings[0]["moisture_pct"] if readings else None,
                   "action": rec["action"], "minutes": rec["minutes"],
                   "reason": rec["reason"], "alerts": open_alerts})
    con.close()
    return result


@mcp.tool(description="Show the reasoning steps for the zone recommendation. zone_id must match a real zone.")
def explain_recommendation(zone_id: int):
    con = _con()
    zone = next((z for z in db.get_zones(con) if z["id"] == zone_id), None)
    if zone is None:
        con.close()
        return {"error": f"zone {zone_id} not found"}
    crop = db.get_crop(con, zone["crop"])
    readings = db.get_recent_readings(con, zone_id, 24)
    now_iso = clock.to_iso(clock.now())
    rec = engine.decide(zone, crop, readings, None, now_iso,
                        clock.local_minutes(clock.now(), _cfg["location"]["utc_offset_minutes"]), _cfg)
    result = _base(con)
    result.update({"action": rec["action"], "reason": rec["reason"], "trace": rec["trace"]})
    con.close()
    return result


@mcp.tool(description="Get current weather status and cache age.")
def get_weather_outlook():
    con = _con()
    key = f"{round(_cfg['location']['latitude'],2)},{round(_cfg['location']['longitude'],2)}"
    cached = db.get_weather_cache(con, key)
    result = _base(con)
    if cached is None:
        result["status"] = "OFFLINE"
        con.close()
        return result
    age_h = (clock.from_iso(clock.to_iso(clock.now())) - clock.from_iso(cached["fetched_at"])).total_seconds() / 3600.0
    data = json.loads(cached["payload"])
    cur = data.get("current", {})
    result.update({"status": "LIVE" if age_h < 0.5 else "CACHED", "cache_age_hours": round(age_h, 1),
                   "temperature_c": cur.get("temperature_2m"), "wind_kmh": cur.get("wind_speed_10m"),
                   "precip_current_mm": cur.get("precipitation")})
    con.close()
    return result


@mcp.tool(description="Water use report per zone. days must be between 1 and 30.")
def water_report(days: int = 7):
    if not 1 <= days <= 30:
        return {"error": "days must be between 1 and 30"}
    con = _con()
    result = _base(con)
    result["assumption"] = "Baseline assumes 30 min irrigation per day per zone at zone flow rate."
    result["zones"] = reports.zone_report(con, db.get_zones(con), _cfg, days)
    con.close()
    return result


@mcp.tool(description="Rain skip ledger with verdicts. limit must be between 1 and 50.")
def rain_check_ledger(limit: int = 10):
    if not 1 <= limit <= 50:
        return {"error": "limit must be between 1 and 50"}
    con = _con()
    rows = con.execute("SELECT * FROM skips ORDER BY decided_at DESC LIMIT ?", (limit,)).fetchall()
    result = _base(con)
    result["summary"] = ledger.ledger_summary(con)
    result["skips"] = [dict(r) for r in rows]
    con.close()
    return result


@mcp.tool(description="Request a valve action. action is open or close, minutes between 1 and max_request_minutes. Requires allow_valve_requests=true.")
def request_valve(zone_id: int, action: str, minutes: int = 30):
    if not _cfg["mcp"]["allow_valve_requests"]:
        return {"error": "valve requests disabled (allow_valve_requests=false in config)"}
    if action not in ("open", "close"):
        return {"error": "action must be open or close"}
    max_min = _cfg["mcp"]["max_request_minutes"]
    if not 1 <= minutes <= max_min:
        return {"error": f"minutes must be between 1 and {max_min}"}
    wcon = db.open_db(_db_path, query_only=False)
    zone = next((z for z in db.get_zones(wcon) if z["id"] == zone_id), None)
    if zone is None:
        wcon.close()
        return {"error": f"zone {zone_id} not found"}
    ts = clock.to_iso(clock.now())
    cur = wcon.execute("INSERT INTO valve_requests (zone_id,action,minutes,requested_at) VALUES (?,?,?,?)",
                       (zone_id, action, minutes, ts))
    wcon.commit()
    req_id = cur.lastrowid
    wcon.close()
    con = _con()
    running, _, _ = _ctrl(con)
    con.close()
    return {"request_id": req_id, "status": "pending", "controller_running": running}


@mcp.resource("sinchai://crops")
def crops_resource():
    con = _con()
    rows = con.execute("SELECT * FROM crops").fetchall()
    con.close()
    return json.dumps([dict(r) for r in rows])


@mcp.resource("sinchai://zones/{zone_id}")
def zone_resource(zone_id: int):
    con = _con()
    zone = next((z for z in db.get_zones(con) if z["id"] == zone_id), None)
    con.close()
    return json.dumps(dict(zone) if zone else {"error": f"zone {zone_id} not found"})


@mcp.resource("sinchai://alerts/recent")
def alerts_resource():
    con = _con()
    rows = db.get_open_alerts(con)
    con.close()
    return json.dumps([dict(r) for r in rows])


@mcp.prompt()
def daily_farm_briefing():
    return """Call list_zones, get_weather_outlook, and get_zone_status for each zone.
Write a short plain-language brief for a farmer: what needs watering, what can wait, and why.
State the data age. If data is simulated, say so. Do not invent numbers."""


def main():
    global _cfg, _db_path
    _db_path = os.environ.get("SINCHAI_DB", "data/sinchai.db")
    _cfg = config.get_config(_db_path)
    mcp.run(transport="stdio")


if __name__ == "__main__":
    main()

File: src/mcp-server.js
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import { z } from "zod";
import { openDb, getZones, getCrop, getRecentReadings, getActiveAlerts, getControllerRow } from "./db.js";
import { getConfig } from "./config.js";
import { decide } from "./engine.js";
import { zoneReport } from "./reports.js";
import { ledgerSummary } from "./ledger.js";
import { now, toIso, fromIso, localMinutes } from "./clock.js";

function getBase(db) {
  const row = getControllerRow(db);
  const nowDt = now();
  if (!row) return { as_of: toIso(nowDt), data_age_minutes: null, controller_running: false, simulated: true };
  const running = (nowDt - fromIso(row.heartbeat_at)) / 1000.0 < 15;
  const age = row.clock_now ? Math.round(((nowDt - fromIso(row.clock_now)) / 60000.0) * 10) / 10 : null;
  return { as_of: row.clock_now, data_age_minutes: age, controller_running: running, simulated: row.source === "sim" || row.demo === 1 };
}

function getZoneRec(db, zoneId, cfg) {
  const zone = getZones(db).find(z => z.id === zoneId);
  if (!zone) return null;
  const readings = getRecentReadings(db, zoneId, 24);
  const crop = getCrop(db, zone.crop);
  const nowDt = now();
  const rec = decide(zone, crop, readings, null, toIso(nowDt), localMinutes(nowDt, cfg.location.utc_offset_minutes), cfg);
  return { zone, readings, rec };
}

export function createMcpServer(dbPath) {
  const db = openDb(dbPath, "src/schema.sql");
  const cfg = getConfig(dbPath);
  const server = new McpServer({ name: "sinchai", version: "1.0.0" });

  server.tool("list_zones", {}, async () => {
    const base = getBase(db);
    const zones = getZones(db).map(z => {
      const r = getRecentReadings(db, z.id, 1)[0];
      const age = r ? Math.round(((now() - fromIso(r.ts)) / 60000.0) * 10) / 10 : null;
      return { id: z.id, name: z.name, crop: z.crop, flow_mm_hr: z.flow_mm_hr, irrigation: z.irrigation, moisture_pct: r?.moisture_pct ?? null, reading_age_minutes: age, valve_open: Boolean(z.valve_open) };
    });
    return { content: [{ type: "text", text: JSON.stringify({ ...base, zones }, null, 2) }] };
  });

  server.tool("get_zone_status", { zone_id: z.number() }, async ({ zone_id }) => {
    const res = getZoneRec(db, zone_id, cfg);
    if (!res) return { content: [{ type: "text", text: JSON.stringify({ error: `zone ${zone_id} not found` }) }] };
    const { zone, readings, rec } = res;
    const alerts = getActiveAlerts(db).filter(a => a.zone_id === zone_id).map(a => ({ kind: a.kind, message: a.message }));
    return { content: [{ type: "text", text: JSON.stringify({ ...getBase(db), zone: zone.name, crop: zone.crop, moisture_pct: readings[0]?.moisture_pct ?? null, action: rec.action, minutes: rec.minutes, reason: rec.reason, alerts }, null, 2) }] };
  });

  server.tool("explain_recommendation", { zone_id: z.number() }, async ({ zone_id }) => {
    const res = getZoneRec(db, zone_id, cfg);
    if (!res) return { content: [{ type: "text", text: JSON.stringify({ error: `zone ${zone_id} not found` }) }] };
    return { content: [{ type: "text", text: JSON.stringify({ ...getBase(db), action: res.rec.action, reason: res.rec.reason, trace: res.rec.trace }, null, 2) }] };
  });

  server.tool("get_weather_outlook", {}, async () => {
    const cached = db.prepare("SELECT * FROM weather_cache WHERE key = 'forecast'").get();
    if (!cached) return { content: [{ type: "text", text: JSON.stringify({ ...getBase(db), status: "OFFLINE" }) }] };
    const ageH = (now() - fromIso(cached.fetched_at)) / 3600000.0;
    const cur = JSON.parse(cached.payload)?.current || {};
    return { content: [{ type: "text", text: JSON.stringify({ ...getBase(db), status: ageH < 0.5 ? "LIVE" : "CACHED", cache_age_hours: Math.round(ageH * 10) / 10, temperature_c: cur.temperature_2m, wind_kmh: cur.wind_speed_10m, precip_current_mm: cur.precipitation }, null, 2) }] };
  });

  server.tool("water_report", { days: z.number().optional().default(7) }, async ({ days }) => {
    if (days < 1 || days > 30) return { content: [{ type: "text", text: JSON.stringify({ error: "days must be between 1 and 30" }) }] };
    const zList = zoneReport(db, getZones(db), cfg, days);
    return { content: [{ type: "text", text: JSON.stringify({ ...getBase(db), assumption: "Baseline assumes 30 min irrigation per day per zone at zone flow rate.", zones: zList }, null, 2) }] };
  });

  server.tool("rain_check_ledger", { limit: z.number().optional().default(10) }, async ({ limit }) => {
    if (limit < 1 || limit > 50) return { content: [{ type: "text", text: JSON.stringify({ error: "limit must be between 1 and 50" }) }] };
    const skips = db.prepare("SELECT * FROM skips ORDER BY decided_at DESC LIMIT ?").all(limit);
    return { content: [{ type: "text", text: JSON.stringify({ ...getBase(db), summary: ledgerSummary(db), skips }, null, 2) }] };
  });

  server.tool("request_valve", { zone_id: z.number(), action: z.enum(["open", "close"]), minutes: z.number().optional().default(30) }, async ({ zone_id, action, minutes }) => {
    if (!cfg.mcp?.allow_valve_requests) return { content: [{ type: "text", text: JSON.stringify({ error: "valve requests disabled (allow_valve_requests=false in config)" }) }] };
    if (minutes < 1 || minutes > (cfg.mcp?.max_request_minutes || 60)) return { content: [{ type: "text", text: JSON.stringify({ error: "invalid minutes" }) }] };
    const zone = getZones(db).find(z => z.id === zone_id);
    if (!zone) return { content: [{ type: "text", text: JSON.stringify({ error: `zone ${zone_id} not found` }) }] };
    const info = db.prepare("INSERT INTO valve_requests (zone_id, action, minutes, requested_at) VALUES (?, ?, ?, ?)").run(zone_id, action, minutes, toIso(now()));
    return { content: [{ type: "text", text: JSON.stringify({ request_id: Number(info.lastInsertRowid), status: "pending", controller_running: getBase(db).controller_running }) }] };
  });

  return server;
}

export async function runMcpServer(dbPath) {
  const server = createMcpServer(dbPath || process.env.SINCHAI_DB || "data/sinchai.db");
  const transport = new StdioServerTransport();
  await server.connect(transport);
}

if (process.argv[1] && process.argv[1].endsWith("mcp-server.js")) {
  runMcpServer(process.env.SINCHAI_DB || "data/sinchai.db").catch(err => {
    console.error("MCP Server Error:", err);
    process.exit(1);
  });
}
